applications/ShapeOptimizationApplication/custom_utilities/response_functions/am_boundary_roughness_angle_response_function_utility.cpp

```diff
@@ -0,0 +1,124 @@
+// ==============================================================================
+//  KratosShapeOptimizationApplication
+//
+//  License:         BSD License
+//                   license: ShapeOptimizationApplication/license.txt
+//
+//  Main authors:    Reza Najian Asl
+//
+// ==============================================================================
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "custom_utilities/response_functions/am_boundary_roughness_angle_response_function_utility.h"
+#include "shape_optimization_application.h"
+#include "utilities/variable_utils.h"
+
+// ==============================================================================
+
+namespace Kratos
+{
+
+AMBoundaryRoughnessAngleResponseFunctionUtility::AMBoundaryRoughnessAngleResponseFunctionUtility(ModelPart& rModelPart, Parameters ResponseSettings)
+	: mrModelPart(rModelPart)
+{
+	const std::size_t domain_size = mrModelPart.GetProcessInfo()[DOMAIN_SIZE];
+	KRATOS_ERROR_IF(domain_size != 3) << "AMBoundaryRoughnessAngleResponseFunctionUtility can only be used on 3D geometries!" << std::endl;
+
+	mMainDirection = ResponseSettings["main_direction"].GetVector();
+	const double direction_norm = norm_2(mMainDirection);
+	KRATOS_ERROR_IF(direction_norm < std::numeric_limits<double>::epsilon()) << "AMBoundaryRoughnessAngleResponseFunctionUtility: 'main_direction' vector norm is 0!" << std::endl;
+	mMainDirection /= direction_norm;
+
+	const double max_angle = ResponseSettings["max_angle"].GetDouble();
+	mCosMaxAngle = std::cos(max_angle * Globals::Pi / 180);
+
+	const std::string gradient_mode = ResponseSettings["gradient_mode"].GetString();
+	if (gradient_mode == "finite_differencing")
+	{
+		mDelta = ResponseSettings["step_size"].GetDouble();
+	}
+	else
+		KRATOS_ERROR << "Specified gradient_mode '" << gradient_mode << "' not recognized. The only option is: finite_differencing" << std::endl;
+}
+
+double AMBoundaryRoughnessAngleResponseFunctionUtility::CalculateValue()
+{
+	KRATOS_TRY;
+
+	const double value = block_for_each<SumReduction<double>>(mrModelPart.Conditions(), [&](Condition& rCond) {
+		return CalculateConditionValue(rCond);});
+
+	mValue = value;
+
+	return mValue;
+
+	KRATOS_CATCH("");
+}
+
+void AMBoundaryRoughnessAngleResponseFunctionUtility::CalculateGradient()
+{
+	KRATOS_TRY;
+	// First gradients are initialized
+	VariableUtils().SetHistoricalVariableToZero(SHAPE_SENSITIVITY, mrModelPart.Nodes());
+
+	for (auto& cond_i : mrModelPart.Conditions()){
```

Comment:
Check [here](https://github.com/KratosMultiphysics/Kratos/blob/90c385c74e8ab56ccacce6bd2eccd36c9f819575/applications/RomApplication/custom_strategies/custom_schemes/modal_derivative_scheme.hpp#L408-L458) if you want to do this in parallel as well

(basically you need to lock all the nodes of the Condition)